File: Problems_and_Solutions/days-10 to 19/day-18/c++/solution 1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* slowPointer=head;
        ListNode* fastPointer=head;
        int count=1;
        while(fastPointer!=NULL)
        {
            fastPointer=fastPointer->next;
            if(++count&1)
                slowPointer=slowPointer->next;
        }
        return slowPointer;
    }
};

File: Problems_and_Solutions/days-10 to 19/day-18/Java/Solution1.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
    }
}

class Solution1 {
    public static ListNode middleNode1(ListNode head) {
        ListNode curr = head;
        int count = 0;

        while(curr != null){
            count++;
            curr = curr.next;
        }

        curr = head;
        for(int i=0; i<count/2; i++){
            curr = curr.next;
        }

        return curr;
    }

    public static ListNode middleNode2(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    public static void printList(ListNode head) {
        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = new ListNode(4);

        ListNode rev = middleNode1(head);
        printList(rev);
    }
}
